### PR TITLE
Don't lose python/name tags values in mkdocs.yml

### LINKED ISSUE #7461: Special !! values are reset to null in mkdocs.yml
## Details

As reported [here](https://github.com/readthedocs/readthedocs.org/issues/6889#issuecomment-638502825) (the issue is now stale, not sure if you'd prefer the discussion to continue over there), values like `!!python/name:module.name` in `mkdocs.yml` are dropped during the build process. They end up being `null` in the final `mkdocs.yml`, which in turn breaks build in various places (mainly plugins).

* Build URL (if applicable): https://readthedocs.org/projects/copier/builds/11817460/

## Expected Result

These values should not be reset to `null` as it breaks the build. More details in the linked issue.

## Actual Result

Special values starting with `!!python/name:` in mkdocs are reset to null, and it breaks the build.

### PR DESCRIPTION
**Pull request by pawamoy**
Close #7461

### CODE REVIEW

**Comment by eeintech** [comment]
@pawamoy If that fix work, we owe you a beer! :beer:

**Comment by pawamoy** [comment]
I'd gladly have a beer with you @eeintech :smile: Living in Strasbourg, France by chance :wink: ?

@stsewd I can squash the commits of course, once you or someone else reviewed the PR :slightly_smiling_face:

**Comment by eeintech** [comment]
> I'd gladly have a beer with you @eeintech  Living in Strasbourg, France by chance  ?

Not in France at the moment, maybe one day when I visit friends in Alsace :wink:

File: readthedocs/config/tests/test_yaml_loader.py
Comment (humitos) [inline readthedocs/config/tests/test_yaml_loader.py:15]: Maybe a test for `!!python/name:module.other.function` is useful as well.

File: readthedocs/config/tests/test_yaml_loader.py
Comment (humitos) [inline readthedocs/config/tests/test_yaml_loader.py:20]: We shouldn't `.pop` the item here and update the `expected` instead.

File: readthedocs/doc_builder/backends/mkdocs.py
Comment (humitos) [inline readthedocs/doc_builder/backends/mkdocs.py:344]:
```suggestion
    Borrowed from https://stackoverflow.com/a/57121993
    Issue https://github.com/readthedocs/readthedocs.org/issues/7461
```

Remove dots to make the link clickable.

File: readthedocs/doc_builder/backends/mkdocs.py
Comment (humitos) [inline readthedocs/doc_builder/backends/mkdocs.py:350]: Shouldn't we need the same for others like: `construct_python_module`, `construct_python_object` and maybe others here as well?

https://github.com/yaml/pyyaml/blob/538b5c93f7d5dee40322893c1e524e94a4f8bbde/lib3/yaml/constructor.py#L572

File: readthedocs/rtd_tests/tests/test_doc_builder.py
Comment (humitos) [inline readthedocs/rtd_tests/tests/test_doc_builder.py:366]: Maybe we want to check for our `SafeDumper` here to be sure, right?

**Review by humitos — CHANGES_REQUESTED** [review]
Thanks!!! Looks like a solid solution IMO. We are leaving as-is these values between load/dump conversion without touching/analyze/execute them. So, I'm :+1: and I think it's safe to merge it.

I left some comments related to the tests and some other small things, but I think we are in a good direction here.

Also, I'd super appreciate if you can create an example in our test-builds repository (https://github.com/readthedocs/test-builds) that uses MkDocs with this scenario so I can test this locally and in production in an easy way.

@stsewd what are your thoughts here?

File: readthedocs/config/tests/test_yaml_loader.py
Comment (pawamoy) [reply readthedocs/config/tests/test_yaml_loader.py:20]: This is what I did originally, but then the object in `expected` must be an instance of `ProxyPythonName`, and when comparing it with the loaded one, it's false, because they are not the same object. This could be fixed by adding an `__eq__` method on the `ProxyPythonName` class, but I felt like it was adding useless code just for the test to pass. However if you think this is the way to go, I'll happily add this `__eq__` method :slightly_smiling_face: Maybe there is another solution though?

File: readthedocs/doc_builder/backends/mkdocs.py
Comment (pawamoy) [reply readthedocs/doc_builder/backends/mkdocs.py:350]: If you think we should immediately add support for all `python/*` tags, then yes. I only added logic for the `python/name` one since it's the only one I use, and the only one that is mentioned in the original issue https://github.com/readthedocs/readthedocs.org/issues/6889.

Just tell me and I'll update the PR to add support for the other tags!

File: readthedocs/rtd_tests/tests/test_doc_builder.py
Comment (pawamoy) [reply readthedocs/rtd_tests/tests/test_doc_builder.py:366]: Indeed, good catch.

File: readthedocs/config/tests/test_yaml_loader.py
Comment (stsewd) [reply readthedocs/config/tests/test_yaml_loader.py:20]: I think we should move the assert to be just after the pop

**Comment by pawamoy** [comment]
@humitos I created a branch to add a test-build: https://github.com/pawamoy/test-builds/tree/mkdocs-python-tags, but I'm not sure how to push that branch into the main repo without merging into master.

File: readthedocs/config/tests/test_yaml_loader.py
Comment (pawamoy) [reply readthedocs/config/tests/test_yaml_loader.py]: Done.

File: readthedocs/config/tests/test_yaml_loader.py
Comment (pawamoy) [reply readthedocs/config/tests/test_yaml_loader.py]: Done. Reorganized the assert statements and added a comment to explain why we pop.

File: readthedocs/doc_builder/backends/mkdocs.py
Comment (stsewd) [reply readthedocs/doc_builder/backends/mkdocs.py:350]: I think it depends on if it's common to have those... But I'd say it doesn't hurt to add them now

**Review by stsewd — APPROVED** [review]

File: readthedocs/doc_builder/backends/mkdocs.py
Comment (humitos) [inline readthedocs/doc_builder/backends/mkdocs.py:386]: I think it may be good to have a test for this one as well to be sure that we are not transforming the YAML and it looks exactly the same after `yaml_dump_safely`, right?

**Comment by humitos** [comment]
> @humitos I created a branch to add a test-build: [https://github.com/pawamoy/test-builds/tree/mkdocs-python-tags](https://github.com/pawamoy/test-builds/tree/mkdocs-python-tags?rgh-link-date=2020-09-28T15%3A59%3A22Z), but I'm not sure how to push that branch into the main repo without merging into master.

Great, thanks! I created a new branch from `master` with the same name you used and opened a PR to merge your branch there.

File: readthedocs/doc_builder/backends/mkdocs.py
Comment (pawamoy) [reply readthedocs/doc_builder/backends/mkdocs.py:350]: There are 16 more Python tags:

tag | python
-- | --
!!python/none | None
!!python/bool | bool
!!python/bytes | (bytes in Python 3)
!!python/str | str (str in Python 3)
!!python/unicode | unicode (str in Python 3)
!!python/int | int
!!python/long | long (int in Python 3)
!!python/float | float
!!python/complex | complex
!!python/list | list
!!python/tuple | tuple
!!python/dict | dict
!!python/name:module.name | module.name
!!python/module:package.module | package.module
!!python/object:module.cls | module.cls instance
!!python/object/new:module.cls | module.cls instance
!!python/object/apply:module.f | value of f(...)

Taken from https://pyyaml.org/wiki/PyYAMLDocumentation#yaml-tags-and-python-types.

Honestly I'm a bit reluctant to add all these, especially at once. Some of these tags might be more complex to support (more than one suffix/value to store in a proxy object, multiple ways to write them, etc.). I think it would be best to add them one by one as people ask for them. We now know how, so it could be quick PRs.

File: readthedocs/doc_builder/backends/mkdocs.py
Comment (pawamoy) [reply readthedocs/doc_builder/backends/mkdocs.py:386]: True! I eventually added the `__eq__` method to `ProxyPythonName` because it made the test super easy to write :rofl: 

```python
def test_yaml_dump_safely():
    data = yaml_load_safely(content)
    assert yaml_load_safely(yaml_dump_safely(data)) == data
```

> exactly the same after yaml_dump_safely

It's not the case, the indentation can change (we don't have indentation here though), and an empty string is added at the end of `python/name` tags (see https://github.com/readthedocs/readthedocs.org/issues/7461#issuecomment-697234615), which is correct and expected, but different from the original input.

File: readthedocs/doc_builder/backends/mkdocs.py
Comment (humitos) [reply readthedocs/doc_builder/backends/mkdocs.py:353]: I think it makes sense to add only what's required now to make MkDocs with plugins to work :+1:

File: readthedocs/doc_builder/backends/mkdocs.py
Comment (humitos) [reply readthedocs/doc_builder/backends/mkdocs.py:389]: As far as the data is the same, I'm OK with that.

**Review by humitos — APPROVED** [review]
I don't know why `eslint` check is failing.

Other than that, I'm :+1: with these changes :100: :tada:

Thanks a lot for your PR!

**Comment by pawamoy** [comment]
> I don't know why eslint check is failing.

Indeed, weird, this PR does not touch JS files :confused: Maybe there was a new release of the tool?

**Comment by stsewd** [comment]
> > I don't know why eslint check is failing.
> 
> Indeed, weird, this PR does not touch JS files confused Maybe there was a new release of the tool?

The problem is fixed in master, you can rebase or merge master into this branch to make the tests pass

**Comment by pawamoy** [comment]
Squashed and rebased on master :slightly_smiling_face:

**Comment by pawamoy** [comment]
Meh, build timeout :cry:

**Comment by eeintech** [comment]
@humitos I'm curious to know what's the usual timing between when a PR is approved and when deployed to your build system? Really eager to be able to use this fix 😃

**Comment by humitos** [comment]
Around one week after it's merged

**Comment by eeintech** [comment]
@humitos And how long after PR is approved would it be merged? I see some PR from last year (2019) still not merged and it concerns me...

**Comment by pawamoy** [comment]
Oh, maybe they were waiting for me to rebase or something, I see there is a conflict. Let me do that.

**Comment by humitos** [comment]
Hi @eeintech! We do our best to merge and deploy PRs. Unfortunately, our time is not enough. I'm targeting this PR to be deployed next week. Let's see if we can make it happen!

**Comment by eeintech** [comment]
@humitos I understand, thanks a lot for your hard work, I hope you can make it happen too!

**Comment by pawamoy** [comment]
I'd still like to see this merged! I'll update each time the bot tries to stale it :smiling_imp:

**Comment by stsewd** [comment]
Not sure where we ended up with a decision about this. /cc @readthedocs/core

**Comment by eeintech** [comment]
@humitos Feels like this PR merge has been dragging...

**Comment by pawamoy** [comment]
@eeintech just keep unstaling and I'm sure good things will happen :grin:

**Review by ericholscher — APPROVED** [review]
We've discussed this internally, and think this is an OK solution in the short term. In the longer term, we'd really love to not do any YAML parsing/dumping, and instead find a better way to integrate with mkdocs via an extension, or some other way that isn't so error-prone.

The main reason we are doing this currently is to add some small things to the config file. We should research and find another way of handling this, probably talking with the mkdocs team to ensure it will be supported going forward.

I'm 👍  on merging this, but we should include a `# TODO:` in the code linking to https://github.com/readthedocs/readthedocs.org/issues/7844

**Comment by eeintech** [comment]
Yay! :clap:

**Comment by steinwelberg** [comment]
I was wondering, is this change already deployed?

**Comment by stsewd** [comment]
@steinwelberg no, sorry, we are going to migrate our platform to AWS, so deploys are frozen till that https://blog.readthedocs.com/aws-migration/

**Comment by steinwelberg** [comment]
Thanks! 

Now I come to wonder, The blog post mentions that you are migrating the community version to AWS. I am actually a business user. I guess this change is also going to be available for business users, right? I guess, that will also be done after the migration to AWS?

hehe, lots of assumptions, I hope you can clarify 😄 .

**Comment by stsewd** [comment]
@steinwelberg we usually deploy both sites together, so .com deploys are frozen as well. And for clarification, .com is already on AWS, we are just going to migrate .org

**Comment by steinwelberg** [comment]
Great, I'll just have to be a bit more patient then 😉 .

**Comment by skchronicles** [comment]
Thank you for the update @steinwelberg,

I am also running into the same emoji issue, and I am looking forward to the new changes. Thank you for taking the time to look into this issue 😄  and I hope your migration goes smoothly and godspeed 💯 🎉  !

**Comment by stsewd** [comment]
@steinwelberg @skchronicles we just deployed the new version in .org and .com, let us know if your problems are solved now.

**Comment by skchronicles** [comment]
@stsewd, I can confirm that the issue was fixed and that my builds are now passing 💯 🎉 💯 🎉 💯 🎉 💯 .
Thank you again! I appreciate it.

**Comment by pawamoy** [comment]
Thank you for your time dear maintainers ❤️

**Pull request merged**
